### PR TITLE
Add unit tests pipelineparam

### PR DESCRIPTION
**Pull request by gaoning777**
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/975)

### CODE REVIEW

File: sdk/python/kfp/dsl/_metadata.py [hunk 1]
```diff
@@ -46,24 +46,36 @@ def to_dict_or_str(self):
...
-      type_meta.name, type_meta.properties = list(json.items())[0]
+    if isinstance(payload, dict):
+      if not _check_valid_type_dict(payload):
+        raise ValueError(payload + ' is not a valid type string')
+      type_meta.name, type_meta.properties = list(payload.items())[0]
       # Convert possible OrderedDict to dict
       type_meta.properties = dict(type_meta.properties)
-    elif isinstance(json, str):
-      type_meta.name = json
+    elif isinstance(payload, str):
```
Comment (hongye-sun) [inline sdk/python/kfp/dsl/_metadata.py]: Do you need to have another else to throw if the payload is not dict or str? I am worry that the ast might convert to unexpected types and we will lose the debug information here.

File: sdk/python/kfp/compiler/compiler.py [hunk 2]
```diff
@@ -219,16 +219,12 @@ def _process_args(self, raw_args, argument_inputs):
     processed_args = list(map(str, raw_args))
     for i, _ in enumerate(processed_args):
       # unsanitized_argument_inputs stores a dict: string of sanitized param -> string of unsanitized param
-      matches = []
-      matches += _match_serialized_pipelineparam(str(processed_args[i]))
+      param_tuples = []
+      param_tuples += _match_serialized_pipelineparam(str(processed_args[i]))
```
Comment (hongye-sun) [inline sdk/python/kfp/compiler/compiler.py]: Thanks for the fix. The code looks much better now.

File: sdk/python/kfp/compiler/compiler.py [hunk 3]
```diff
@@ -219,16 +219,12 @@ def _process_args(self, raw_args, argument_inputs):
...
       unsanitized_argument_inputs = {}
-      for x in list(set(matches)):
-        if len(x) == 3 or (len(x) == 4 and x[3] == ''):
-          sanitized_str = str(dsl.PipelineParam(K8sHelper.sanitize_k8s_name(x[1]), K8sHelper.sanitize_k8s_name(x[0]), x[2]))
-          unsanitized_argument_inputs[sanitized_str] = str(dsl.PipelineParam(x[1], x[0], x[2]))
-        elif len(x) == 4:
-          sanitized_str = str(dsl.PipelineParam(K8sHelper.sanitize_k8s_name(x[1]), K8sHelper.sanitize_k8s_name(x[0]), x[2], TypeMeta.from_dict_or_str(x[3])))
-          unsanitized_argument_inputs[sanitized_str] = str(dsl.PipelineParam(x[1], x[0], x[2], TypeMeta.from_dict_or_str(x[3])))
+      for param_tuple in list(set(param_tuples)):
```
Comment (hongye-sun) [inline sdk/python/kfp/compiler/compiler.py]: Could you confirm if named tuple can dedup the same values with different instances here?

File: sdk/python/tests/dsl/pipeline_param_tests.py [hunk 4]
```diff
@@ -51,4 +52,14 @@ def test_extract_pipelineparam(self):
...
\ No newline at end of file
+  def test_extract_pipelineparam_with_types(self):
+    p1 = PipelineParam(name='param1', op_name='op1', param_type=TypeMeta(name='customized_type_a', properties={'property_a': 'value_a'}))
+    p2 = PipelineParam(name='param2', param_type=TypeMeta(name='customized_type_b'))
+    p3 = PipelineParam(name='param3', value='value3', param_type=TypeMeta(name='customized_type_c', properties={'property_c': 'value_c'}))
+    stuff_chars = ' between '
+    payload = str(p1) + stuff_chars + str(p2) + stuff_chars + str(p3)
+    params = _extract_pipelineparams(payload)
+    self.assertListEqual([p1, p2, p3], params)
+    payload = [str(p1) + stuff_chars + str(p2), str(p2) + stuff_chars + str(p3)]
```
Comment (hongye-sun) [inline sdk/python/tests/dsl/pipeline_param_tests.py]: nit: Adding a comment here that this test verifies that the method will dedup parameters if it appears in is several places.

File: sdk/python/kfp/compiler/compiler.py (hunk 3 above)
Comment (hongye-sun) [reply sdk/python/kfp/compiler/compiler.py]: Found the test that covers this scenario. Thanks.

File: sdk/python/kfp/dsl/_metadata.py (hunk 1 above)
Comment (gaoning777) [reply sdk/python/kfp/dsl/_metadata.py]: Good idea. Done.

File: sdk/python/kfp/compiler/compiler.py (hunk 2 above)
Comment (gaoning777) [reply sdk/python/kfp/compiler/compiler.py]: Thank you for pointing that out. I think it is important to address these concerns to keep the maintenance cost low.

File: sdk/python/tests/dsl/pipeline_param_tests.py (hunk 4 above)
Comment (gaoning777) [reply sdk/python/tests/dsl/pipeline_param_tests.py]: Good point. Done

**Comment by gaoning777** [comment]
/test kubeflow-pipeline-e2e-test
--

**Pull request merged**